paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -458,11 +458,30 @@ def can_kill(
             killable_capacity = round(sum([slave.instance_weight for slave in filtered_slaves]), 2)
             amount_to_decrease = round(delta * -1, 2)
             if amount_to_decrease > killable_capacity:
-                self.log.error(
-                    "Didn't find enough candidates to kill. This shouldn't happen so let's not kill anything! "
-                    "Amount wanted to decrease: %s, killable capacity: %s" % (amount_to_decrease, killable_capacity),
+                instance_ids_in_mesos = {
+                    slave.instance_id for slave in filtered_slaves
+                }
+                instance_ids_not_in_mesos = [
+                    instance['InstanceId'] for instance in self.instances
+                    if instance['InstanceId'] not in instance_ids_in_mesos
+                ]
+                self.log.warning(
+                    "Didn't find enough candidates to kill. This may mean that "
+                    "some instances have not yet joined the cluster. Since we "
+                    "are scaling down, we will kill the instances in this "
+                    "pool that have not joined.\n"
+                    "Desired decrease in capacity: %s, killable capacity: %s\n"
+                    "Setting capacity to %s and killing instances %s before "
+                    "proceeding scale down" % (
+                        amount_to_decrease,
+                        killable_capacity,
+                        killable_capacity,
+                        instance_ids_not_in_mesos,
+                    ),
                 )
-                return
+                self.set_capacity(killable_capacity)
+                self.terminate_instances(instance_ids_not_in_mesos)
```

Comment:
Yeah, we'll end up here if we're trying to scale down and there are instances in the SFR/ASG that haven't registered themselves in Mesos.  We (drmorr, hliu) figured that since we're trying to scale down, we don't really want them anyway (apparently Clusterman makes the same decision).

`amount_to_decrease` is calculated from the number of instances in the SFR/ASG, while `killable capacity` is the total capacity of instances from this SFR/ASG that are registered in Mesos, so if `amount_to_decrease > killable_capacity` we know there are instances in the SFR/ASG that aren't in Mesos, and `self.set_capacity(killable_capacity)` should reduce the SFR/ASG's capacity to just what's currently in Mesos, preventing it from scaling back up after we kill the instances.

Also `SpotAutoscaler.set_capacity()` automatically rounds the capacity down and converts it to an int (see [here](https://github.com/Yelp/paasta/blob/64d72c2a0b51bc6edbe98192150a98783b5115c0/paasta_tools/autoscaling/autoscaling_cluster_lib.py#L843))

Lemme know if that doesn't make sense!

Reply:
Thanks for the explanation :-)

That all sounds sensible to me. Ship it!